text/1026-net-module.rst

```diff
@@ -0,0 +1,208 @@
+       named only to: multi str,
+       named only subject: str,
+       named only text: optional str,
+       named only html: optional str,
+   ) -> std::net::SmtpResponse
+
+ResponseState enum
+------------------
+
+Since network requests are asynchronous, the response state will be
+represented by an enum:
+
+.. code-block:: edgeql
+
+   scalar type std::net::ResponseState extending std::enums<Pending, InProgress, Complete, Failed>;
+
+Response Object Types
+---------------------
+
+The HTTP response object will be an internal type with the following
+structure:
+
+.. code-block:: edgeql
+
+   type std::net::HttpResponse {
+       required state: std::net::ResponseState;
+       required created_at: datetime;
+
+       status: int16;
+       headers: json;
+       body: bytes;
+   }
+
+
+The SMTP response object will be an internal type with the following
+structure:
+
+.. code-block:: edgeql
+
+   type std::net::SmtpResponse {
+       required state: std::net::ResponseState;
+       required created_at: datetime;
+
+       reply_code: int16;
+       reply_message: str;
+   }
+
+Implementation Details
+----------------------
+
+1. Requests will be stored in a queue table in the database.
+2. A Rust process will handle sending the requests.
+3. Each protocol (HTTP, SMTP) will have its own queue and pool of worker
+   processes.
+4. Simple retry logic will be implemented for failed requests.
+5. URLs will initially be represented as plain strings, with the
+   possibility of adding type-checked URL support in the future.
+
+Examples
+========
+
+HTTP Request
+------------
+
+.. code:: edgeql
+
+   with
+       payload := '{"key": "value"}',
+       response := (
+           select std::net::http_request(
+               'POST',
+               'https://api.example.com/webhook',
+               body := payload,
+               headers := to_json('{"Content-Type": "application/json"}'),
+           )
+       )
+   select response {
+       id,
+       state,
+       created_at,
+   };
+
+SMTP Send
+---------
+
+.. code:: edgeql
+
+   with
+       html_body := '<html><body><p>Hello, this is a test email.</p></body></html>',
+       text_body := 'Hello, this is a test email.',
+       response := (
+           select std::net::smtp_send(
+               'smtp://smtp.example.com:587',
+               from := 'sender@example.com',
+               to := {'recipient1@example.com', 'recipient2@example.com'},
+               subject := 'Test Email',
+               html := html_body,
+               text := text_body
+           )
+       )
+   select response {
+       id,
+       state,
+       created_at,
+   };
+
+Backwards Compatibility
+=======================
+
+This RFC introduces new functionality and does not affect existing
+features. There are no backwards compatibility issues.
+
+Security Implications
+======================
+
+1. URL Validation: Implement proper URL validation to prevent malicious
+   URL inputs.
```

Comment:
I'd rephrase -- "URL validation" implies that we somehow verify the correctness of a URL right before the request is made. What we should do instead is having a std::URL type that would implement correct construction and quoting/concatenation.